### ISSUE TITLE
MapLibre induces self-intersection in valid polygon

### ISSUE THREAD

**Issue by leofuhrmann** [maplibre/maplibre-gl-js#3156]
**maplibre-gl-js version**: 3.3.1

**browser**: Firefox 118

### Steps to Trigger Behavior

 1. Create vector tile from [example GeoPackage](https://github.com/maplibre/maplibre-gl-js/files/12783412/example.gpkg.zip) using ogr2ogr `ogr2ogr -f MVT -dsco EXTENT=16384 -dsco MINZOOM=15 -dsco MAXZOOM=15 -dsco SIMPLIFICATION=0 tiles example.gpkg`.
 2. Create a tile JSON file.
 3. Serve the necessary files using a web server, e. g. locally using Express.js.
 4. Create a map using MapLibre GL JS, rendering fill and line for the example data's polygon.

### Link to Demonstration

You can see a demonstration here: https://jsbin.com/bajirozelo/edit?html,output

Comment out the layers or remove the comments to hide/show expected/actual behavior.

### Expected Behavior

The polygon is rendered without self-intersection:

![image](https://github.com/maplibre/maplibre-gl-js/assets/126690962/0b805239-3a4d-4645-96c5-aed1fd196bff)

Note: This polygon is created using geoJSON, which was converted from the vector tile (using `ogr2ogr -t_srs EPSG:4326 -f GeoJSON -lco RFC7946=YES example.geojson tiles/15/17270/10767.pbf`).

### Actual Behavior

The polygon is rendered with an induced self-intersection (bottom-right), creating visual artifacts:

![image](https://github.com/maplibre/maplibre-gl-js/assets/126690962/47da32cb-0927-4504-b1cd-35e0f9ea66d1)

**Comment by HarelM** [comment]
Is this a duplicate of #1830?

**Comment by leofuhrmann** [comment]
> Is this a duplicate of #1830?

No, because the issue you mentioned addresses the rendering of already invalid polygons, if I understood correctly. The current issue deals with a valid polygon (without intersections), that seemingly becomes invalid during rendering.

**Comment by HarelM** [comment]
It might be that the conversion accuracy is problematic. I would be surprised if the rendering is causing the invalid polygon.
Did you check that the vector tile data is correct and not intersecting itself?
You can use the https://github.com/mapbox/vector-tile-js package to inspect a single tile data and extract the polygon, I would start there.
If the data is not intersecting itself there then the issue might be with the transition of the data to the worker/GPU, but I would be surprised if this is the case.
Nevertheless, everything is possible :-)

**Comment by HarelM** [comment]
I would try and simplify this issue first, make a polygon with 4-6 points, try even a geojson as example to see if it reproduces there and then follow the code that uses the shaders programs to see what is sent to the GPU in the vertices.
It might be a limitation of the data sent to the GPU, IDK...

**Comment by leofuhrmann** [comment]
I have updated my first comment and tried my best to include a minimal example to reduce the problem. It uses a valid GeoJSON, from which a vector tile is created. The vector tile is converted back to GeoJSON to check the validity of the tile's content. Both are displayed in the demonstration (https://jsbin.com/xadehayoru/edit?html,output). One should clearly see the intersection in the rendered tile.

So far, I couldn't look into the rendering process, but as the geojson is rendered correctly, this seems to be connected to the "unpacking" or rendering of vector tiles.

**Comment by HarelM** [comment]
The geojson's accuracy is a lot higher than the tile's accuracy because the tile max zoom is 15 while the geojson max zoom is unlimited (24 in this example).
When I add the following property to the geojson source I get the same results:
`maxzoom: 15`.

```js
map.addSource('geojson', {
            'type': 'geojson',
            'maxzoom': 15, // this is the interesting part!
            'data': { // I just inlined the data for readability
                "type": "FeatureCollection",
                "name": "example",
                "features": [
                    { 
                        "type": "Feature", 
                        "properties": { }, 
                        "geometry": { 
                            "type": "MultiPolygon", 
                            "coordinates": [[[
                                
                                [ 9.740717634558678, 52.376506310061373 ], 
                                [ 9.740708246827126, 52.376498941724329 ], 
                                [ 9.740707576274872, 52.376504672653247 ], 
                                [ 9.740713611245154, 52.376503444597105 ], 
                                [ 9.740708917379379, 52.376505491357321 ], 
                                [ 9.740717634558678, 52.376506310061373 ] 
                            ] ] ] 
                        } 
                    }
                ]
            }
        });
```

This is how this looks in the map:

<img width="481" alt="image" src="https://github.com/maplibre/maplibre-gl-js/assets/3269297/d1483e8b-f013-4ba6-9838-ceeb5b0e7935">

I'm not saying there isn't a bug here in terms of moving data to the GPU, but looking at a feature in zoom 24 while having a tile that has accuracy of zoom 15 is asking for accuracy issues, doesn't it?

**Comment by leofuhrmann** [comment]
> looking at a feature in zoom 24 while having a tile that has accuracy of zoom 15 is asking for accuracy issues, doesn't it?

You're right, this kind of "overzoom" is surely questionable. However, the intersection being created can lead to artifacts visible in smaller zoom levels, if the problematic vertices are part of a bigger polygon:

The problematic part is located at the bottom right of the following polygon. It leads to a visual artifact (seemingly "cutting" the polygon) being visible from zoom 15.5:

<img width="86" alt="image" src="https://github.com/maplibre/maplibre-gl-js/assets/126690962/409420ed-b7ac-4cec-8d6a-738c0151d1f9">

So, at some point coordinates might be translated without checking again if the resulting polygon is still "valid". Such a check and resulting measures are necessary in my opinion. If anyone has an idea, I'll happily try to contribute.

**Comment by HarelM** [comment]
I started looking into it, but it requires time.
Your best bet would be to start digging.
Since this can be reproduced with a geojson source of 6 points I would start there.
It might be a 32 bit limitation, IDK...
The geojson worker source is where I started debugging, but didn't find anything there, so it's probably the uniform array that is passed to the GPU.
There's a PR which increased the "extent" and an issue that talks about overzoom, feel free to poke around those.

**Comment by CurtisFenner** [comment]
I just ran into this issue with a polygon as small as 7 vertices.

It seems that maplibre-gl is incorrectly splitting up a polygon _when it spans more than 3 tiles_.

Add `map.showTileBoundaries = true;` makes this abundantly clear:

Large  polygon where it is immediately obvious:

![image](https://github.com/maplibre/maplibre-gl-js/assets/3289892/6ba9a3cf-f841-4a00-99ba-0abfa91e94b8)

![image](https://github.com/maplibre/maplibre-gl-js/assets/3289892/2103ca5c-aa77-4358-a99b-0e5e1f20cbe5)

Here is a very small polygon (a hexagon) where the problem occurs:

![image](https://github.com/maplibre/maplibre-gl-js/assets/3289892/4d2f18c9-3e7e-4c77-bb64-6bab173e93a7)

![image](https://github.com/maplibre/maplibre-gl-js/assets/3289892/b7b23f65-658a-4f4a-9bf4-f0518bb0f474)

I don't think there's anything special about the polygon itself, but here it is for reproduction:

```json
[
    [
        [
            139.5046234808873,
            35.605187046891146
        ],
        [
            139.50577340782908,
            35.60762779416374
        ],
        [
            139.50939725344256,
            35.60762779416374
        ],
        [
            139.510485622726,
            35.60507129239125
        ],
        [
            139.51066891696067,
            35.602724732337514
        ],
        [
            139.5044093046226,
            35.60262350327767
        ],
        [
            139.5046234808873,
            35.605187046891146
        ]
    ]
]
```

**Comment by HarelM** [comment]
I'm not sure it's the same issue... I would recommend opening a different issue about this issue.

**Comment by CommanderStorm** [comment]
Not reproducible any more because the s3 link used is not usable anymore..
I tried reproducing with just the images this seems to be a bit hard.

We can reopen if there is a reproduction or in a fresh issue ^^
